File: problems/0291-word-pattern-ii/solution.cpp
class Solution {
private:
    unordered_map<char, string> char_to_str;
    unordered_map<string, char> str_to_char;
    
    bool backtrack(const string& pattern, int p_idx, const string& s, int s_idx) {
        // Base case: both pattern and string are fully matched.
        if (p_idx == pattern.length() && s_idx == s.length()) {
            return true;
        }
        
        // One finished but not the other.
        if (p_idx == pattern.length() || s_idx == s.length()) {
            return false;
        }
        
        // Current pattern character.
        char c = pattern[p_idx];
        
        // If char already has a mapping, verify it matches.
        if (char_to_str.find(c) != char_to_str.end()) {
            string mapped_str = char_to_str[c];
            // Check if s has enough chars and matches the mapping.
            if (s_idx + mapped_str.length() <= s.length() && 
                s.substr(s_idx, mapped_str.length()) == mapped_str) {
                return backtrack(pattern, p_idx + 1, s, s_idx + mapped_str.length());
            }
            return false;
        }
        
        // Try all possible substrings for this char.
        for (int end = s_idx + 1; end <= s.length(); end++) {
            string substring = s.substr(s_idx, end - s_idx);
            
            // Skip if substring already mapped to another char.
            if (str_to_char.find(substring) != str_to_char.end()) {
                continue;
            }
            
            // Create new mapping.
            char_to_str[c] = substring;
            str_to_char[substring] = c;
            
            // Recurse with this mapping.
            if (backtrack(pattern, p_idx + 1, s, end)) {
                return true;
            }
            
            // Backtrack: remove mapping.
            char_to_str.erase(c);
            str_to_char.erase(substring);
        }
        
        return false;
    }
    
public:
    bool wordPatternMatch(string pattern, string s) {
        /**
         * Problem: 0291 - Word Pattern II
         * 
         * Check if string s can be segmented to match the pattern where each
         * character in pattern maps to a substring in s with bijection.
         * Uses backtracking to try all possible segmentations.
         * 
         * Approach:
         * - Use DFS with backtracking to try all substring mappings
         * - Maintain char->substring and substring->char mappings
         * - For each pattern char, try all possible substring lengths
         * - Backtrack if mapping leads to contradiction
         * 
         * Time complexity: O(n^m) worst case where n = len(s), m = len(pattern).
         * Space complexity: O(m) for recursion stack and mappings.
         */
        return backtrack(pattern, 0, s, 0);
    }
};

File: problems/0291-word-pattern-ii/solution.py
class Solution:
    def wordPatternMatch(self, pattern: str, s: str) -> bool:
        """
        Problem: 0291 - Word Pattern II
        
        Check if string s can be segmented to match the pattern where each
        character in pattern maps to a substring in s with bijection.
        Uses backtracking to try all possible segmentations.
        
        Approach:
        - Use DFS with backtracking to try all substring mappings
        - Maintain char->substring and substring->char mappings
        - For each pattern char, try all possible substring lengths
        - Backtrack if mapping leads to contradiction
        
        Time complexity: O(n^m) worst case where n = len(s), m = len(pattern).
        Space complexity: O(m) for recursion stack and mappings.
        """
        def backtrack(p_idx, s_idx):
            # Base case: both pattern and string are fully matched.
            if p_idx == len(pattern) and s_idx == len(s):
                return True
            
            # One finished but not the other.
            if p_idx == len(pattern) or s_idx == len(s):
                return False
            
            # Current pattern character.
            char = pattern[p_idx]
            
            # If char already has a mapping, verify it matches.
            if char in char_to_str:
                mapped_str = char_to_str[char]
                # Check if s has enough chars and matches the mapping.
                if (s_idx + len(mapped_str) <= len(s) and 
                    s[s_idx:s_idx + len(mapped_str)] == mapped_str):
                    return backtrack(p_idx + 1, s_idx + len(mapped_str))
                return False
            
            # Try all possible substrings for this char.
            for end in range(s_idx + 1, len(s) + 1):
                substring = s[s_idx:end]
                
                # Skip if substring already mapped to another char.
                if substring in str_to_char:
                    continue
                
                # Create new mapping.
                char_to_str[char] = substring
                str_to_char[substring] = char
                
                # Recurse with this mapping.
                if backtrack(p_idx + 1, end):
                    return True
                
                # Backtrack: remove mapping.
                del char_to_str[char]
                del str_to_char[substring]
            
            return False
        
        # Mappings for bijection.
        char_to_str = {}
        str_to_char = {}
        
        return backtrack(0, 0)
